File: app/src/main/java/com/mkulesh/onpc/iscp/messages/BroadcastResponseMsg.java
/*
 * Enhanced Music Controller
 * Copyright (C) 2018-2022 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

package com.mkulesh.onpc.iscp.messages;

import com.mkulesh.onpc.iscp.EISCPMessage;
import com.mkulesh.onpc.iscp.ISCPMessage;
import com.mkulesh.onpc.utils.Utils;

import java.net.InetAddress;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/*
 * Broadcast Response Message
 *
 * !cECNnnnnnn/ppppp/dd/iiiiiiiiiiii:
 * c: device category
 * nnnnnnn: model name of device
 * ppppp: ISCP port number
 * dd: destination area of device
 * iiiiiiiiiiii: Identifier
 * /: Separator
 */
public class BroadcastResponseMsg extends ISCPMessage
{
    public final static int DCP_PORT = 23;
    private final static String CODE = "ECN";

    private String model = null;
    private String destinationArea = null;
    private String identifier = null;
    private String alias = null;

    public BroadcastResponseMsg(InetAddress hostAddress, EISCPMessage raw) throws Exception
    {
        super(raw);
        host = hostAddress.getHostAddress();
        String[] tokens = data.split("/");
        if (tokens.length > 0)
        {
            model = tokens[0];
        }
        if (tokens.length > 1)
        {
            port = Integer.parseInt(tokens[1], 10);
        }
        if (tokens.length > 2)
        {
            destinationArea = tokens[2];
        }
        if (tokens.length > 3)
        {
            identifier = tokens[3];
        }
    }

    public BroadcastResponseMsg(BroadcastResponseMsg other)
    {
        super(other);
        this.model = other.model;
        this.port = other.port;
        this.destinationArea = other.destinationArea;
        this.identifier = other.identifier;
        this.alias = other.alias;
    }

    public BroadcastResponseMsg(@NonNull final String host, final int port,
                                @NonNull final String alias, @Nullable final String identifier)
    {
        super(0, null);
        this.host = host;
        this.port = port;
        this.alias = alias;
        this.identifier = identifier;
        // all other fields still be null
    }

    public BroadcastResponseMsg(@NonNull final String host, final int port,
                                @NonNull final String model)
    {
        super(0, null);
        this.host = host;
        this.port = port;
        this.model = model;
        // all other fields still be null
    }

    @NonNull
    @Override
    public String toString()
    {
        return CODE + "[" + data
                + "; HOST=" + getHostAndPort()
                + "; " + getProtoType()
                + (model != null ? "; MODEL=" + model : "")
                + (destinationArea != null ? "; DST=" + destinationArea : "")
                + (identifier != null ? "; ID=" + identifier : "")
                + (alias != null ? "; ALIAS=" + alias : "") + "]";
    }

    @NonNull
    public String getDescription()
    {
        final String d = alias != null ? alias : (model != null ? model : "unknown");
        return getHost() + "/" + d;
    }

    @NonNull
    public String getIdentifier()
    {
        return identifier == null ? "" : identifier;
    }

    @Nullable
    public String getAlias()
    {
        return alias;
    }

    public Utils.ProtoType getProtoType()
    {
        return port == DCP_PORT ? Utils.ProtoType.DCP : Utils.ProtoType.ISCP;
    }
}

File: lib/iscp/messages/BroadcastResponseMsg.dart
/*
 * Enhanced Music Controller
 * Copyright (C) 2019-2025 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

import 'dart:io';

import "../../utils/Convert.dart";
import "../ConnectionIf.dart";
import "../EISCPMessage.dart";
import "../ISCPMessage.dart";
import "../upnp/DeviceDescription.dart";

/*
 * Broadcast Response Message
 *
 * !cECNnnnnnn/ppppp/dd/iiiiiiiiiiii:
 * c: device category
 * nnnnnnn: model name of device
 * ppppp: ISCP port number
 * dd: destination area of device
 * iiiiiiiiiiii: Identifier
 * /: Separator
 */

typedef OnDeviceFound = void Function(BroadcastResponseMsg msg);

class BroadcastResponseMsg extends ISCPMessage with ProtoTypeMix
{
    static const String CODE = "ECN";

    String? _model;
    String? _destinationArea;

    // Used to store optional MAC address
    String? _identifier;

    // Used to store optional favourite name
    String? _alias;

    // Used to store optional friendly name
    String? _friendlyName;

    // Optional UPnP device description message
    DeviceDescription? _upnpDescription;

    BroadcastResponseMsg(InternetAddress hostAddress, EISCPMessage raw) : super(CODE, raw)
    {
        setHost(hostAddress.address);
        final List<String> tokens = getData.split("/");
        if (tokens.isNotEmpty)
        {
            _model = tokens[0];
        }
        if (tokens.length > 1)
        {
            setPort(ISCPMessage.nonNullInteger(tokens[1], 10, 0));
        }
        if (tokens.length > 2)
        {
            _destinationArea = tokens[2];
        }
        if (tokens.length > 3)
        {
            _identifier = _trim(tokens[3]);
        }
        setProtoType(ProtoType.ISCP);
        // _alias and _friendlyName still be null - not presented in ISCP
    }

    BroadcastResponseMsg.alias(final String host, final String port, final String alias, final String? identifier) : super.output(CODE, "")
    {
        setHost(host);
        setPort(ISCPMessage.nonNullInteger(port, 10, 0));
        this._identifier = identifier;
        this._alias = alias;
        setProtoType(getPort == DCP_PORT ? ProtoType.DCP : ProtoType.ISCP);
        // all other fields still be null
    }

    BroadcastResponseMsg.connection(ConnectionIf connection, final String? alias, final String? identifier) : super.output(CODE, "")
    {
        setHostAndPort(connection);
        this._identifier = identifier;
        this._alias = alias;
        setProtoType(getPort == DCP_PORT ? ProtoType.DCP : ProtoType.ISCP);
        // all other fields still be null
    }

    BroadcastResponseMsg.ssdp(InternetAddress hostAddress, ProtoType p, final DeviceDescription d) : super.output(CODE, "")
    {
        setHost(hostAddress.address);
        setPort(p == ProtoType.ISCP? ISCP_PORT : DCP_PORT);
        setProtoType(p);
        this._upnpDescription = d;
        this._model = d.modelName;
        this._friendlyName = d.friendlyName;
        // all other fields still be null
    }

    String _trim(String token)
    {
        String res = "";
        for (int i = 0; i < token.length; i++)
        {
            if (token.codeUnitAt(i) == EISCPMessage.EOM)
            {
                break;
            }
            res += token[i];
        }
        return res;
    }

    @override
    String toString()
    => super.toString() + "[HOST=" + getHostAndPort
            + "; " + Convert.enumToString(protoType)
            + (_model != null ? "; MODEL=" + _model! : "")
            + (_destinationArea != null ? "; DST=" + _destinationArea! : "")
            + (_identifier != null ? "; ID=" + _identifier! : "")
            + (_alias != null ? "; ALIAS=" + _alias! : "")
            + (_friendlyName != null ? "; FRIENDLY_NAME=" + _friendlyName! : "") + "]";

    String getDescription(bool useFavoriteName, bool useFriendlyName)
    {
        if (useFavoriteName && _alias != null)
        {
            return _alias!;
        }
        if (useFriendlyName && _friendlyName != null)
        {
            return _friendlyName!;
        }
        return getHost + "/" + (_model != null ? _model! : "unknown");
    }

    String get getIdentifier
    => _identifier ?? "";

    String? get alias
    => _alias;

    String? get friendlyName
    => _friendlyName;

    set friendlyName(String? value)
    {
        _friendlyName = value;
    }

    DeviceDescription? get upnpDescription
    => _upnpDescription;
}
